### ISSUE TITLE
Suggest ptr::null_mut() when .is_null() is better

### ISSUE THREAD

**Issue by kpp** [rust-lang/rust-clippy#4174]
I got this warning:

```
warning: `0 as *mut _` detected. Consider using `ptr::null_mut()`
    --> ocl-core/src/functions.rs:3605:18
     |
3605 |     if ext_fn == 0 as *mut c_void {
     |                  ^^^^^^^^^^^^^^^^
     |
```

I replaced it as it was suggested:

```
if ext_fn == ptr::null_mut() {
```

However clippy gave me another warning:

```
warning: Comparing with null is better expressed by the .is_null() method
    --> ocl-core/src/functions.rs:3605:8
     |
3605 |     if ext_fn == ptr::null_mut() {
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
```

**Comment by flip1995** [comment]
Is `if ext_fn.is_null() { ... }` correct code or is the second suggestion wrong?

If the second suggestion is correct, this behavior is intended. Sometimes a lint suggests an improvement, which leads to code, that can still be improved. As long as there is no suggestion loop, this is fine.

To avoid this and directly suggest to use `.is_null` would mean, that Clippy would need to 
- check the code
- apply suggestions
- check the new code
- ...

This would be a fixpoint iteration, which would take a **reeeeaaaally** long time (Running the compiler frontend and every lint multiple times).

**Comment by kpp** [comment]
> If the second suggestion is correct, this behavior is intended

The second suggestion is correct.

**Comment by flip1995** [comment]
In this case I'm going ahead and close this issue, since this is intended behavior, which can't really be improved, since these lints are and should be independent from each other.

https://github.com/rust-lang/rust-clippy/blob/42f96b2558abd0d235aee87ce4b3a167d5df0663/clippy_lints/src/ptr.rs#L67-L69

https://github.com/rust-lang/rust-clippy/blob/b0ec33f661a53737bc87972cbbab2b3218e66476/clippy_lints/src/misc.rs#L208-L210

**Comment by kpp** [comment]
Why not create another lint like CMP_ZERO_PTR?

**Comment by flip1995** [comment]
This would be unnecessary code duplication in Clippy, because the two lints already cover this case. When now adding a new lint (or enhancing the CMP_NULL lint to detect this), two warnings would be emitted:
1. comparison with zero pointer
2. cast of zero pointer

This could be avoided by checking if the cast is in a comparison. But this would add even more code to Clippy. The cost-benefit ratio improvement is too high for this IMHO.

When fixing errors/lint findings in your code, you prabably almost always have to do more than one pass. So "two step linting" shouldn't be that bad.

**Comment by kpp** [comment]
OK, thanks
